File: Sorting/Quick_sort/quick_sort.py
def partition(arr, low, high):

    pivot = arr[low]
    i = low
    j = high

    while(i < j):

        while i <= high-1 and arr[i] <= pivot:
            i +=1
        while j >= low+1 and arr[j] > pivot:
            j -= 1

        if i < j:
            arr[i], arr[j] = arr[j], arr[i]
    arr[low], arr[j] = arr[j], arr[low]

    return j


def quick_sort(arr, low, high):

    if low < high:
        pivot_index = partition(arr, low, high)

        quick_sort(arr, low, pivot_index-1)
        quick_sort(arr, pivot_index+1, high)


# Driver code
arr = [10, 7, 8, 9, 1, 5]


# Call quicksort
quick_sort(arr, 0, len(arr) - 1)

# Print sorted array
print(*arr)

File: Sorting/Quick_sort/quick_sort.cpp
#include <bits/stdc++.h>
using namespace std;

int partition(vector<int>& arr, int low, int high) {
    int pivot = arr[low];

    int i = low + 1;
    int j = high;

    while (i <= j) {

        // Check the boundary BEFORE accessing arr[i]
        while (i <= high && arr[i] <= pivot) {
            i++;
        }

        // Check the boundary BEFORE accessing arr[j]
        while (j >= low + 1 && arr[j] > pivot) {
            j--;
        }

        if (i < j) {
            swap(arr[i], arr[j]);
        }
    }

    swap(arr[low], arr[j]);
    return j;
}

void quickSort(vector<int>& arr, int low, int high) {
    if (low < high) {
        int pivotIndex = partition(arr, low, high);

        quickSort(arr, low, pivotIndex - 1);
        quickSort(arr, pivotIndex + 1, high);
    }
}

int main() {
    
    vector<int> arr = {10, 7, 8, 9, 1, 5};

    quickSort(arr, 0, arr.size() - 1);

    for (int num : arr)
        cout << num << " ";

    return 0;
}